internal/buff/write.go

```diff
@@ -0,0 +1,167 @@
+func NewWriter() *Writer {
+	w := &Writer{}
+	w.buf = w.alocatedMemory[:0]
+	return w
+}
+
+// Send writes buffered data to conn.
+func (w *Writer) Send(conn io.Writer) (err error) {
+	if w.msgPos != 0 {
+		panic("cannot send: the previous message is not finished")
+	}
+
+	if len(w.buf) == 0 {
+		panic("cannot send: no data")
+	}
+
+	var n int
+	n, err = conn.Write(w.buf)
+	if err != nil {
+		return err
+	}
+
+	if n != len(w.buf) {
+		panic("not all data was sent")
+	}
+
+	w.buf = w.alocatedMemory[:0]
+	return err
+}
+
+// PushUint8 writes a uint8 to the buffer.
+func (w *Writer) PushUint8(val uint8) {
+	w.buf = append(w.buf, val)
+}
+
+// PushUint16 writes a uint16 to the buffer.
+func (w *Writer) PushUint16(val uint16) {
+	n := len(w.buf)
+	w.buf = append(w.buf, 0, 0)
+	binary.BigEndian.PutUint16(w.buf[n:n+2], val)
+}
+
+// PushUint32 writes a uint32 to the buffer.
+func (w *Writer) PushUint32(val uint32) {
+	n := len(w.buf)
+	w.buf = append(w.buf, 0, 0, 0, 0)
+	binary.BigEndian.PutUint32(w.buf[n:n+4], val)
+}
+
+// PushUint64 writes a uint64 to the buffer.
+func (w *Writer) PushUint64(val uint64) {
+	n := len(w.buf)
+	w.buf = append(w.buf, 0, 0, 0, 0, 0, 0, 0, 0)
+	binary.BigEndian.PutUint64(w.buf[n:n+8], val)
+}
+
+// PushUUID writes a types.UUID to the buffer.
+func (w *Writer) PushUUID(val types.UUID) {
+	w.buf = append(w.buf, val[:]...)
+}
+
+// PushBytes writes a []byte to the buffer.
+func (w *Writer) PushBytes(val []byte) {
+	w.PushUint32(uint32(len(val)))
+	w.buf = append(w.buf, val...)
+}
+
+// PushString writes a string to the buffer.
+func (w *Writer) PushString(val string) {
+	w.PushBytes([]byte(val))
+}
+
+// BeginBytes allocates space for `data_length` in the buffer.
+// May be called multiple times to create nested bytes blocks.
+// Calling EndBytes once for each BeginBytes call is required
+// before ending a message.
+// BeginBytes panics if BeginMessage was not called first.
+func (w *Writer) BeginBytes() {
+	if w.msgPos == 0 {
+		panic("cannot begin bytes: no current message")
+	}
+
+	n := len(w.buf)
+	w.buf = append(w.buf, 0, 0, 0, 0)
+	w.bytePos = append(w.bytePos, n)
+}
+
+// EndBytes sets the `data_length` allocated by BeginBytes
+// to the number of bytes that were written since the last BeginBytes call.
+// EndBytes panics if BeginBytes was not called first.
+func (w *Writer) EndBytes() {
+	n := len(w.bytePos)
+	if n < 1 {
+		panic("cannot end bytes: no bytes in progress")
+	}
+
+	pos := w.bytePos[n-1]
+	w.bytePos = w.bytePos[:n-1]
+
+	byteLen := uint32(len(w.buf) - pos - 4)
+	binary.BigEndian.PutUint32(w.buf[pos:], byteLen)
+}
+
+// BeginMessage writes mType to the buffer
+// and allocates space for message length.
+// BeginMessage panics if the EndMessage was not called
+// for the previous message.
+func (w *Writer) BeginMessage(mType uint8) {
+	if w.msgPos != 0 {
+		panic("cannot begin message: the previous message is not finished")
+	}
+
+	w.msgPos = 1 + len(w.buf)
+	w.buf = append(w.buf, mType, 0, 0, 0, 0)
```

Comment:
Is `append` a stdlib function? What will it do when you are at the end of `alocatedMemory`?

Reply:
Yes its stdlib. it will allocate a new underly array that is ~ `2 * len(w.buf)`, copy the data in `w.buf` to the new array followed by the 5 bytes listed and return a new slice into the new array that ends at the last new 0. I think this is similar to the way `list.append()` works in python :thinking: